locust_plugins/transaction_manager.py

```diff
@@ -0,0 +1,238 @@
+    user_count = 0
+    csv_headers = [
+        "start_time",
+        "duration",
+        "transaction_name",
+        "user_count",
+        "success",
+        "failure_message",
+    ]
+
+    def __init__(self):
+        self.transaction_count = 0
+        self.inprogress_transactions = {}
+        # results filename format
+        # fields set for file output
+
+    @classmethod
+    def _command_line_parser(cls, parser):
+        parser.add_argument(
+            "--log_transactions_in_file",
+            help="To log transactions in file rather than using the web ui, set to True",
+            default=False,
+        )
+        cls.is_not_worker = not parser.parse_args().worker
+
+    def start_transaction(self, transaction_name):
+        now = time()
+        transaction = {}
+        transaction["transaction_name"] = transaction_name
+        transaction["start_time"] = now
+        transaction["end_time"] = None
+        transaction["duration"] = None
+        transaction["success"] = None
+        transaction["failure_message"] = None
+        transaction["user_count"] = self.runner.user_count if self.runner else 0
+        self.inprogress_transactions[transaction_name] = transaction
+        return transaction_name
+
+    def end_transaction(self, transaction_name, success=True, failure_message=""):
+        now = time()
+        t = self.inprogress_transactions[transaction_name]
+        t["end_time"] = now
+        start_time = t["start_time"]
+        t["duration"] = round((now - start_time) * 1000)
+        t["success"] = success
+        t["failure_message"] = failure_message
+
+        self.flat_transaction_list.append(
+            f'{datetime.fromtimestamp(t["start_time"]).strftime(self.timestamp_format)}{self.field_delimiter}'
+            f'{round(t["duration"])}{self.field_delimiter}'
+            f'{t["transaction_name"]}{self.field_delimiter}'
+            f'{t["user_count"]}{self.field_delimiter}'
+            f'{t["success"]}{self.field_delimiter}'
+            f'{t["failure_message"]}'
+        )
+
+        if t["transaction_name"] not in self.completed_transactions:
+            self.completed_transactions[t["transaction_name"]] = []
+
+        self.completed_transactions[t["transaction_name"]].append(t)
+
+        del self.inprogress_transactions[transaction_name]
+
+        if len(self.flat_transaction_list) >= self.flush_size and self.log_transactions_in_file and self.is_not_worker:
+            self._flush_to_log()
+
+    @classmethod
+    def _create_results_log(cls):
+        results_file = open(cls.transactions_filename, "w")
+        results_file.write(cls.field_delimiter.join(cls.csv_headers) + cls.row_delimiter)
+        results_file.flush()
+        return results_file
+
+    @classmethod
+    def _flush_to_log(cls):
+        cls.results_file.write(cls.row_delimiter.join(cls.flat_transaction_list) + cls.row_delimiter)
+        cls.results_file.flush()
```

Comment:
For better performance, and data integrity (but most importantly: simplicity in the code), call .write() every time (instead of maintaing a list) and instead rely on the normal file buffering (depends on OS, but supposedly it will be something like 4 or 8KB)

Reply:
Maybe you were inspired by the TimescaleListener to do this, but it only makes sense if there is an overhead to writing the event (there is close to zero overhead in calling .write() on a buffered file, so no need to worry about that)

Reply:
Thanks for the tip - I didn't know this.

In this particular case, a list needs to be maintained because writing to file is only one of three possible outcomes.
The other two are to report to master or use the web_ui endpoint.